src/treasuryDAO.mligo

```diff
@@ -23,22 +23,35 @@ let treasury_DAO_proposal_check (params, extras : propose_params * contract_extr
   let max_xtz_amount = unpack_tez(find_big_map("max_xtz_amount", extras)) in
 
   let required_token_lock = frozen_scale_value * proposal_size + frozen_extra_value in
-  let has_correct_token_lock =
-    (params.frozen_token = required_token_lock) && (proposal_size < max_proposal_size) in
+  let pm = unpack_proposal_metadata(params.proposal_metadata) in
 
-  if has_correct_token_lock then
-    let pm = unpack_proposal_metadata(params.proposal_metadata) in
+  let _ : unit =
+    if (params.frozen_token <> required_token_lock) then
+      ([%Michelson ({| { FAILWITH } |} : (string * string) -> unit)]
+        ("FAIL_PROPOSAL_CHECK", "Incorrect token amounts locked") : unit)
+    else if (proposal_size >= max_proposal_size) then
+       ([%Michelson ({| { FAILWITH } |} : (string * string) -> unit)]
+        ("FAIL_PROPOSAL_CHECK", "Proposal size is bigger than 'max_proposal_size'") : unit)
+    else
+      unit in
 
-    let is_all_transfers_valid (is_valid, transfer_type: bool * transfer_type) =
-      match transfer_type with
-      | Token_transfer_type _tt -> is_valid
-      | Xtz_transfer_type xt ->
-             is_valid && xt.amount <> 0mutez
-          && min_xtz_amount <= xt.amount && xt.amount <= max_xtz_amount
-    in
-      List.fold is_all_transfers_valid pm.transfers true
-  else
-    false
+  let is_all_transfers_valid (acc, transfer_type: unit * transfer_type) =
+    match transfer_type with
+    | Token_transfer_type _tt -> acc
+    | Xtz_transfer_type xt ->
+        if (xt.amount = 0mutez) then
+          ([%Michelson ({| { FAILWITH } |} : (string * string) -> unit)]
+            ("FAIL_PROPOSAL_CHECK", "Xtz transfer amount cannot be 0") : unit)
```

Comment:
A couple of comments about this (in general, it applies to all of these new errors):

1. We should still use short strings for these errors, there is either way a cost associated with them.
2. This change to `FAIL_PROPOSAL_CHECK` needs to be documented in the `Errors` section of the spec document, same for all these new "options" for the right side of the pair (perhaps it's worth making a separate table?)
3. These can be abstracted to a function, which would be more convenient:
   ```ocaml
   let fail_proposal_check (msg : string) : unit =
     ([%Michelson ({| { FAILWITH } |} : (string * string) -> unit)]
       ("FAIL_PROPOSAL_CHECK", msg) : unit)
   ```

Reply:
I've added point 1 and 3 now. 

Related to point 2, since these errors are only related to registry and treasury DAO, do we need to add it to the spec? Or maybe we should add it to `registry.md` and `treasury.md` instead?

Reply:
Yeah about that, I was thinking of separating them somehow:
> perhaps it's worth making a separate table?

I was thinking of making a list of "known errors", but your idea to use `registry.md` and `treasury.md` is better, let's go with that.

Reply:
Yeah about that, I was thinking of separating them somehow:
> perhaps it's worth making a separate table?

I was thinking of making a list of "known errors", but your idea to use `registry.md` and `treasury.md` is better, let's go with that.

Reply:
I've added the errors to them now.